main.go

```diff
@@ -40,59 +40,42 @@ func main() {
 
 	// initialize resolver
 	config := records.SetConfig(*cjson)
-	resolver := resolver.New(version, config)
+	res := resolver.New(version, config)
 	errch := make(chan error)
 
 	// launch DNS server
 	if config.DNSOn {
-		go func() { errch <- <-resolver.LaunchDNS() }()
+		go func() { errch <- <-res.LaunchDNS() }()
 	}
 
 	// launch HTTP server
 	if config.HTTPOn {
-		go func() { errch <- <-resolver.LaunchHTTP() }()
+		go func() { errch <- <-res.LaunchHTTP() }()
 	}
 
-	var newLeader <-chan struct{}
+	var masters <-chan []string
 	var zkErr <-chan error
 	// launch Zookeeper listener
 	if config.Zk != "" {
-		newLeader, zkErr = resolver.LaunchZK(zkInitialDetectionTimeout)
+		masters, zkErr = res.LaunchZK(zkInitialDetectionTimeout)
 		go func() { errch <- <-zkErr }()
 	} else {
 		// uniform behavior when new leader from masters field
-		leader := make(chan struct{}, 1)
-		leader <- struct{}{}
-		newLeader = leader
+		mc := resolver.NewMasterChanger(config.Masters)
+		masters = mc.ChangedChan()
+		go mc.Emit()
 	}
 
-	// generate reload signal; up to 1 reload pending at any time
-	reloadSignal := make(chan struct{}, 1)
-	tryReload := func() {
-		// non-blocking, attempt to queue a reload
-		select {
-		case reloadSignal <- struct{}{}:
-		default:
-		}
-	}
-
-	// periodic loading of DNS state (pull from Master)
-	go func() {
+	reloadTimeout := time.Second * time.Duration(config.RefreshSeconds)
+	for {
 		defer util.HandleCrash()
-		reloadTimeout := time.Second * time.Duration(config.RefreshSeconds)
-		reloadTimer := time.AfterFunc(reloadTimeout, tryReload)
-		for _ = range reloadSignal {
-			resolver.Reload()
-			logging.PrintCurLog()
-			reloadTimer.Reset(reloadTimeout)
-		}
-	}()
 
-	// infinite loop until there is fatal error
-	for {
 		select {
-		case <-newLeader:
-			tryReload()
+		case ms := <-masters:
+			res.Masters(ms)
+			res.Reload()
+		case <-time.After(reloadTimeout):
+			res.Reload()
```

Comment:
Note that I used a new timer at each iteration, but `time.Ticker` is even better :-)